### PR TITLE
Export traces to App Insights via OpenTelemetry

### LINKED ISSUE #1113: Export traces in OpenTelemetry format and then to Azure Monitor
/kind feature
/help

**Describe the solution you'd like**
As a user of Cluster API Provider Azure, I would like to be able to export trace spans in OpenTelemetry format so that they can then be exported to Azure Monitor via the [Azure Monitor Exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/master/exporter/azuremonitorexporter).

Currently, when a developer runs `make tilt-up`, Jaeger is started on http://localhost:8080 and a Jaeger "all-in-one" sidecar is injected to into the controller pod. The CAPZ controller sends traces to the Jaeger sidecar for direct analysis.

The desired solution is to be able to run `make tilt-up` and have https://github.com/open-telemetry/opentelemetry-collector or https://github.com/open-telemetry/opentelemetry-operator setup with the CAPZ controller exporting traces to a collector configured to export to Azure Monitor using the Azure Monitor OpenTelemetry Exporter.

We should also document the components and configuration needed to create a CAPZ cluster with these components installed for non-development scenarios. (@mboersma, this might be an interesting idea for a CLI gesture)

**Anything else you would like to add:**
The first tracing PR #1019 added Jaeger support. It is a good starting point to learn from, but lacks the flexibility of design which comes with using the OpenTelemetry collector. Separating the collector from the exporter / analysis will enable users of CAPZ to have more control over their stack of tools.

### PR DESCRIPTION
**Pull request by mboersma**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Removes jaeger from the `make tilt-up` workflow and replaces it with [opentelemetry-collector](https://github.com/open-telemetry/opentelemetry-collector) (still operating in jaeger-compatible collector mode). Adds the [azuremonitor exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/azuremonitorexporter) component to send CAPZ controller-monitor tracing to App Insights for visualization in the "Transaction Search" function.

**Which issue(s) this PR fixes**:

Fixes #1113
Depends on #1412

**Special notes for your reviewer**:

**TODOs**:

- [x] squashed commits
- [x] includes documentation
- [ ] adds unit tests
- [x] Merge #1412 and rebase
- [x] Testing by others, since `make tilt-up` understandably isn't covered by any test harness
- [x] Documentation on how to install the Helm chart on an existing cluster
- [x] Documentation (with pictures!) on how to set up Azure Insights and retrieve the instrumentation key
- [x] Any other TODOs in the readme.md
- [x] Warn if `AZURE_INSTRUMENTATION_KEY` isn't set, but ensure the cluster continues to function (without tracing)
- [ ] Document a local trace destination -- maybe how to install the Jaeger all-in-one helm chart?
- [ ] Maybe switch to the otlp collector? Or make that a followup PR

**Release note**:

```release-note
Export traces to App Insights via OpenTelemetry
```

### CODE REVIEW

**Comment by mboersma** [comment]
This works, but needs more polishing before it's ready to go:

- [ ] Testing by others, since `make tilt-up` understandably isn't covered by any test harness
- [ ] Documentation on how to install the Helm chart on an existing cluster
- [ ] Documentation on how to set up Azure Insights and retrieve the instrumentation key
- [ ] Any other TODOs in the readme.md
- [ ] Maybe switch to the otlp collector? Or make that a followup PR

File: hack/observability/opentelemetry/readme.md
Comment (CecileRobertMichon) [inline hack/observability/opentelemetry/readme.md:20]: is this a shared value? shouldn't it be the dev's App insights resource key like stated above?

Comment (CecileRobertMichon) [reply hack/observability/opentelemetry/readme.md:20]: I guess that's what that sentence means, my first initial reading of the doc (skipping directly to the copy pastable content) read this as "Use this value" as in the value below, maybe clarify this should be the value mentioned above

**Comment by mboersma** [comment]
/retest

DNS failure, seems unrelated to this change.

File: hack/observability/opentelemetry/readme.md
Comment (mboersma) [reply hack/observability/opentelemetry/readme.md]: I updated this language to be more clear.

**Comment by mboersma** [comment]
This is working and ready for review when time permits.

I have another large set of changes to follow this, which replaces the jaeger plumbing with otlp and exports traces to a local Jaeger all-in-one container, but this PR is big enough already so I'm holding those out.

File: hack/observability/opentelemetry/chart/Chart.yaml
Comment (CecileRobertMichon) [inline hack/observability/opentelemetry/chart/Chart.yaml:7]: is having the maintainers of a Helm chart in the repo common practice?

File: hack/observability/opentelemetry/readme.md
Comment (CecileRobertMichon) [inline hack/observability/opentelemetry/readme.md:1]: what do you think about linking this from the development.md doc so it's more discoverable?

**Review by CecileRobertMichon — COMMENTED** [review]
lgtm

I tried it out locally, took a while but eventually got the traces in App Insights. Thanks for writing great docs!

File: hack/observability/opentelemetry/chart/Chart.yaml
Comment (mboersma) [reply hack/observability/opentelemetry/chart/Chart.yaml:7]: I think it's recommended to have maintainers in the Chart.yaml.

What we did here is fetch the chart and store it as a set of local files, to avoid installing a tilt addon for fetching remote charts, to avoid fetching things over the internet, and to pin at a certain chart version.

File: hack/observability/opentelemetry/readme.md
Comment (mboersma) [reply hack/observability/opentelemetry/readme.md:1]: Good call. I should have updated the "View Telemetry" section in the book regardless, and now I did and linked to the opentelemetry readme.md. (Not squashed yet.)

Comment (mboersma) [reply hack/observability/opentelemetry/readme.md:1]: Also as soon as this merges, I'll open the next PR that actually restores Jaeger all-in-one just to view traces without needing App Insights. So some of the language here will return.

Comment (mboersma) [reply hack/observability/opentelemetry/readme.md:1]: I updated this PR to include the opentelemetry-collector LICENSE file for the Helm chart.

**Review by devigned — APPROVED** [review]
/lgtm

**Comment by devigned** [comment]
@CecileRobertMichon +1. Not sure if we want to wait until after release, so I didn't approve.

**Pull request merged**
